internal/blockchain/validate.go

```diff
@@ -1014,6 +1035,163 @@ func (b *BlockChain) isOldBlockVersionByMajority(header *wire.BlockHeader, block
+	// The ability to determine whether or not the blake3 proof of work agenda
+	// is active is not possible in the general case here because that relies on
+	// additional context that is not available in the positional checks.
+	// However, it is important to check for valid bits in the positional checks
+	// to protect against various forms of malicious behavior.
+	//
+	// Thus, with the exception of the special cases where it is possible to
+	// definitively determine the agenda is active, allow valid difficulty bits
+	// under both difficulty algorithms while rejecting blocks that satisify
+	// neither here in the positional checks and allow the contextual checks
+	// that happen later to ensure the difficulty bits are valid specifically
+	// for the correct difficulty algorithm as determined by the state of the
+	// blake3 proof of work agenda.
+	// -------------------------------------------------------------------------
+
+	// Ensure the difficulty specified in the block header matches the
+	// calculated difficulty using the algorithm defined in DCP0011 when the
+	// blake3 proof of work agenda is always active.
+	if b.isBlake3PowAgendaForcedActive() {
+		blake3Diff := b.calcNextBlake3Diff(prevNode)
+		if header.Bits != blake3Diff {
+			str := fmt.Sprintf("block difficulty of %d is not the expected "+
+				"value of %d (difficulty algorithm: ASERT)", header.Bits,
+				blake3Diff)
+			return ruleError(ErrUnexpectedDifficulty, str)
+		}
+
+		return nil
+	}
+
+	// A cached anchor point for the difficulty algorithm defined in DCP0011 is
+	// set when the contextual checks have determined that the agenda is
+	// actually active.
+	//
+	// Since contextual checks for previous block(s) may or may not have been
+	// done yet, this can't be relied on entirely, however, it is a nice
+	// optimization when it can be used.
+	//
+	// In practice, once the chain is synced, and the agenda is active, the
+	// anchor will be set and thus this can run immediately.
+	cachedAnchor := b.cachedBlake3WorkDiffAnchor.Load()
+	if cachedAnchor != nil && cachedAnchor.IsAncestorOf(prevNode) {
+		blake3Diff := b.calcNextBlake3DiffFromAnchor(prevNode, cachedAnchor)
+		if header.Bits != blake3Diff {
+			str := fmt.Sprintf("block difficulty of %d is not the expected "+
+				"value of %d (difficulty algorithm: ASERT)", header.Bits,
+				blake3Diff)
+			return ruleError(ErrUnexpectedDifficulty, str)
+		}
+
+		return nil
+	}
+
+	// Only the original difficulty algorithm needs to be checked when it is
+	// impossible for the blake 3 proof of work agenda to be active or the block
+	// is not solved for blake 3.
+	isSolvedBlake3 := func(header *wire.BlockHeader) bool {
+		powHash := header.PowHashV2()
+		err := standalone.CheckProofOfWorkHash(&powHash, header.Bits)
+		return err == nil
+	}
+	minBlake3BlockVersion := uint32(10)
+	if !isMainNet(b.chainParams) {
+		minBlake3BlockVersion++
+	}
+	isBlake3PossiblyActive := uint32(header.Version) >= minBlake3BlockVersion
```

Comment:
Looking through this again, I see this could also check for a height > svh+rcai*2 here now since a new check was added above for the always active case meaning the only remaining possible way for the agenda to be active for it to have been voted in which requires voting to be possible (svh), at least one interval of voting (rcai), and one interval of locked in (rcai) to have passed.

Reply:
I've updated this accordingly.